### DISCUSSION TITLE
Would persistent stateful index lists be desirable?

### DISCUSSION THREAD

**Discussion by rraub** [activeadmin/activeadmin#7864] [Help]
For a few of my Active Admin applications I've found that persisting the parameters for the index pages in the session gives the users a more persistent interface.

Heres what I've been doing to hack this in:
https://gist.github.com/rraub/6735432

Not the most elegant solution, but seems to work well. 

What drove me to do this was seeing users use the index filters to narrow down the resource list, then edit one of the results and return back to the index page of the resource list without any of their filters they had just set.

I'm wondering if this is a feature others might want to, or if there are other use cases for this kind of feature.

**Comment by seanlinsley** [comment]
I did something like this over in #1926: [[link](https://github.com/gregbell/active_admin/pull/1926/files#L5R77)]

``` diff
+    def safe_params
+      request.query_parameters.slice *::ActiveAdmin.application.safe_params
+    end
+    helper_method :safe_params
+
+    def default_url_options
+      super.merge safe_params
+    end
```

But that was for a completely different goal

**Comment by seanlinsley** [comment]
But to answer the question, I'd really like to see something like what you describe. I'm just not sure what the scope of this functionality should be.

**Comment by rraub** [comment]
That is similar.. I'm not taking into account what params to persist (just everything except for the controller and action), but I probably should.

Really the best use of this feature is on the index pages, we have some users that always use a particular scope or filter and instead of always leaving a tab open of their search while they work they can now go back and forth and not loose their spot. I can see more use for this if there is some configuration for the interface (like showing/hiding a column). 

Arguably this state info could get persisted in the database to work across multiple user sessions. 

Would you like me to write up a cucumber test or something to demonstrate this more formally?

**Comment by zorab47** [comment]
Maybe provide a previous search link above the filters that could fill in the search options and submit the form?

**Comment by hazah** [comment]
The standard solution is to use a destination parameter in your URLs that can override the default redirection. Ensure that the proper query params are set in that paremeter and you're good to go.

**Comment by rraub** [comment]
Passing parameters intended for future use in the url params links seems like a limited solution, because if you want to retain those params for more than one page you'll need to pass them around all over the place. If its a one page and then back kind of workflow then it could work, but what about between other index pages?

index -> view -> edit -> view -> index  is the typical workflow I've observed for most of my Active Admin applications. Passing the index's query parameters along to all these views seems like a lot of work when the session is a reasonable place to put and retrieve this.

A previous search link, does sound like good idea, that would enable us to store a history of previous searches. Why stop there, what about "saved" or "favorited" searches for each user?

**Comment by hazah** [comment]
@rraub This is not intended to be carried across several requests, so you are correct in your assessment that that kind of workflow will not lend itself to what I suggest. The destination param in my experience is intended for one off situation in your workflow. For instance, you've filtered an index and clicked on 'edit', the param will point to that index filtered and on save you will return to that index filtered. If you click 'show' on the other hand, then on edit, you will NOT get the destination param and the redirection will be back to the show action

**Comment by ball-hayden** [comment]
I'd really like to see this as well - I find it counter-intuitive that when I edit something from an index page, when I get redirected my previous filters / sorts no longer apply.

**Comment by seanlinsley** [comment]
Maybe instead of trying to make the server remember the user's filters, it'd be better to store it with HTML5 Web Storage, and fallback to a cookie for IE < 8. [[ref]](https://developer.mozilla.org/en-US/docs/Web/Guide/API/DOM/Storage)

**Comment by rraub** [comment]
We could store the users filters on the client, however that would potentially require a page reload with the new filters  since the only way to get at the stored values is using javascript. I still think storing them in the session isn't that bad of an idea. 

I would like to find an excuse to use HTML5 Web Storage for something..

**Comment by seanlinsley** [comment]
It wouldn't be that difficult to implement.
1. you'd store filters in relation to the current URL so as not to clash with other index pages
2. on page load, if there aren't any filters already applied but we have filters stored in a cookie (or whatever), then apply them
3. if the user clicks the existing Clear Filters button, this cache gets cleared.

**Comment by ball-hayden** [comment]
Wouldn't that mean an extra request though?
You'd have to load the index page, then reload the page to apply the filters?

**Comment by seanlinsley** [comment]
Hmm, yeah you're right. I guess this would have to be handled by the server.

**Comment by tinynumbers** [comment]
Here's an approach I came up with that saves filters between different renderings of a given resource's index page in AA:

https://gist.github.com/tinynumbers/7235594

This only saves filters, it does not persist the last page you were viewing.

**Comment by rraub** [comment]
@tinynumbers that is a nicer implementation than I've been using. Have you thought about expanding the saved filters to include sort and maybe the current page?

On a related note, does anyone consider modifying the params hash bad practice? You are kind of side stepping strong_parameters, but it is a simple use case and it does provide good feature isolation..

**Comment by seanlinsley** [comment]
I don't see any problem with it. I wouldn't say this is side stepping strong_parameters, because these specific params aren't destructive.

While we might want to store the current sort, I'm not sure it makes a lot of UI sense to keep the current page.

**Comment by tinynumbers** [comment]
I agree with @seanlinsley - I can see cases where saving the current page might actually be annoying (for example when doing a bulk delete, the page you were previously on might not exist anymore).

Saving the sort order would be a pretty simple extension of what I've done for saving the filters.

On a user-experience note, another feature I added to my current  usages of this filter-saving is a CSS class that highlights currently-applied filters. The class is applied via some JS as follows:

``` coffee
  # Add an 'active' class to sidebar filters that were selected when the
  # page was loaded.
  $('#filters_sidebar_section').find('select[id],input[type="text"]').each ->
    unless $(this).attr('id') is ''
      if $(this).val()
        $(this).parent('.filter_form_field').addClass('active')
```

Then I can make the active filter fields pop visually (make them bright green or something), which helps avoid user confusion when coming back to the index page with re-applied filters.

**Comment by seanlinsley** [comment]
Related: https://github.com/gregbell/active_admin/issues/2122

**Comment by MissHenesy** [comment]
I ran into this quandary and I have applied my own special band-aid.

I've used a combination of the active_admin.js.coffee file for some javascript magic, combined with a little session magic on the server side in my Rails file.

My objective:: Persist the user's filters from the index page to the edit page, so that when a record is edited, they can return to their filtered list.

In active_admin.js.coffee::
I manipulate the href values of the "Edit" link on the index page, and the "Cancel" button on the edit page, so that each button retains the query string.

What I achieve::
- I maintain the querystring on the Edit page
- The "Update" button will redirect to itself, ie, the Edit page, with the query string intact
- The "Cancel" button (now renamed "Return to List") will return user to Index page, query string / filters intact

Here is the code I used to achieve the above::

### CODE FROM ACTIVE_ADMIN.JS.COFFEE FILE::

```
#= require active_admin/base
$ ->
  set_action_buttons()

set_action_buttons = ->
  querystring = window.location.search
  linkbutton_edit = $("#active_admin_content").find("a.edit_link")
  if linkbutton_edit.length && querystring
    # Append the querystring to the edit button's href link
    linkbutton_edit.each ->
      href = $(this).attr("href") + querystring
      $(this).attr "href", href
      return
  if get_page_title = "some page title"
    # Check the page title name, because we don't necessarily want to 
    # manipulate *all* of active_admin's cancel buttons ....
    linkbutton_cancel = $("#active_admin_content").find("li.cancel").children("a")
    if linkbutton_cancel.length
     # We have a cancel button - let's change its text from "Cancel" to "Return to List"
      linkbutton_cancel.text("Return to List")

get_page_title = ->
  this_page_class_name = $("body").attr("class")
  if /some_class_text/i.test(this_page_class_name)
    "some page title"
```

### CODE FROM APP/ADMIN/MY_ADMIN.RB FILE

```
# Placement of Submit/Cancel Buttons in the .rb file ... this is within the "form do" routine::
form do |f|
    # form stuff
    f.semantic_errors *f.object.errors.keys
     inputs "Stuff to Update" do
         input :some_field
    end
    panel "Submit Buttons" do
      para "Press cancel to return to the list without saving."
     actions do
        action(:submit)
        if session[:redirect_url]
         # We have a redirect url, let's append the querystring to the 
         # href of our choosing in the cancel button
          qs = session[:redirect_url].partition('?').last
          cancel_link("/admin/my_admin_page?" + qs)
        end
      end
end

# Controller to manipulate Submit button
  controller do
    def update
      super do |success,failure|
        success.html { redirect_to get_redirect_url }
      end
    end
    def get_redirect_url
      if request.env["HTTP_REFERER"].include? "?"
       # Oh good, we have a query string! Let's set our session value 
       # equivalent to the referrer link.
        session[:redirect_url] = request.env["HTTP_REFERER"]
      else
       # Uh oh, no query string. Let's use whatever we put in our session value.
        session[:redirect_url]
      end
    end
  end
```

Nuthin' to it! Probably not the most elegant solution, but it works for me :-)

**Comment by abrambailey** [comment]
I strongly support making this the default behavior, or at least making it settable in an initializer.
